File: trunk/cgs/cgs.js
var N = 1000000;

function playGame() {
	var playerBet = 0;
	var playerBetAmount = 1;
	var numRounds = 0;
	var playerBalance = 0;
	
	while (true) {
		numRounds++;
		playerBalance -= playerBetAmount;
		
		var coin = Math.random() > 0.5 ? 1 : 0;
		if (coin === playerBet) {
			playerBalance += 2 * playerBetAmount;
			break;
		}
		
		playerBetAmount *= 2;
	}
	
	return {'playerBalance': playerBalance, 'numRounds': numRounds};
}

function simulation() {
	var gameResult = playGame();
	var minPlayerBal = gameResult.playerBalance;
	var maxPlayerBal = minPlayerBal;
	var accumNumRounds = gameResult.numRounds;
	
	for (var i = 1; i < N; i++) {
		gameResult = playGame();
		
		accumNumRounds += gameResult.numRounds;
		if (gameResult.playerBalance < minPlayerBal) {
			minPlayerBal = gameResult.playerBalance;
		}
		if (gameResult.playerBalance > maxPlayerBal) {
			maxPlayerBal = gameResult.playerBalance;
		}
	}
	
	return {
		'minPlayerBal': minPlayerBal,
		'maxPlayerBal': maxPlayerBal,
		'accumNumRounds': accumNumRounds
	}
}

function $(id) {
	return document.getElementById(id);
}

function onStartup() {
	$('launch').onclick = onLaunch;
}

function onLaunch() {
	var simRes = simulation();
	$('minPlayerBal').innerHTML = simRes.minPlayerBal;
	$('maxPlayerBal').innerHTML = simRes.maxPlayerBal;
	$('avgNumRounds').innerHTML = simRes.accumNumRounds / N;
}

onStartup();

File: trunk/cgs/cgs.cpp
#include <climits>
#include <cstdio>
#include <cstdlib>

const int N = 1000000;

void play_game(int *player_balance, int *num_rounds)
{
    int player_bet = 0;
    int player_bet_amount = 1;
    while (true)
    {
       (*num_rounds)++;
       (*player_balance) -= player_bet_amount;
       int coin = rand() % 2;
       if (coin == player_bet)
       {
           (*player_balance) += 2 * player_bet_amount;
           break;
       }
       player_bet_amount *= 2;
    }
}

void simulation(void)
{
    int min_player_bal = INT_MAX;
    int max_player_bal = INT_MIN;
    int accum_num_rounds = 0;

    for (int i = 0; i < N; i++)
    {
        int player_balance = 0;
        int num_rounds = 0;

        play_game(&player_balance, &num_rounds);
        
        accum_num_rounds += num_rounds;
        if (player_balance < min_player_bal)
            min_player_bal = player_balance;
        if (player_balance > max_player_bal)
            max_player_bal = player_balance;
    }

    printf("Best result: %d\n", max_player_bal);
    printf("Worst result: %d\n", min_player_bal);
    printf("Average number of coin flips before winning: %f\n",
        (double)accum_num_rounds / N);
}

int main(void)
{
    srand(0);
    simulation();
    return 0;
}
